src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequestFactory.java

```diff
@@ -59,16 +61,36 @@ private static SqlRequest parseSqlRequestFromPayload(RestRequest restRequest) {
         JSONObject jsonContent;
         try {
             jsonContent = new JSONObject(content);
+            if (jsonContent.has(SQL_CURSOR_FIELD_NAME)) {
+                return new SqlRequest(jsonContent.getString(SQL_CURSOR_FIELD_NAME));
+            }
         } catch (JSONException e) {
             throw new IllegalArgumentException("Failed to parse request payload", e);
         }
         String sql = jsonContent.getString(SQL_FIELD_NAME);
+
         if (jsonContent.has(PARAM_FIELD_NAME)) { // is a PreparedStatement
             JSONArray paramArray = jsonContent.getJSONArray(PARAM_FIELD_NAME);
             List<PreparedStatementRequest.PreparedStatementParameter> parameters = parseParameters(paramArray);
             return new PreparedStatementRequest(sql, jsonContent, parameters);
         }
-        return new SqlRequest(sql, jsonContent);
+        return new SqlRequest(sql, validateAndGetFetchSize(jsonContent), jsonContent);
+    }
+
+
+    private static Integer validateAndGetFetchSize(JSONObject jsonContent) {
+        Integer fetchSize = null;
```

Comment:
null or 0?

Reply:
+1

Reply:
`null` to imply that `fetch_size` was not passed in query. In `DefaultQueryAction`, if `fetch_size == null`, we will use the `fetch_size` value from cluster settings.

`fetch_size == 0` will imply falling back to non-paginated behavior.

Reply:
I suggest to use Optional instead of null value in this case.

Reply:
Done.